### ISSUE TITLE
apt-get update/upgrade/install downloads packages at unacceptably slow speeds.

### ISSUE THREAD

**Issue by RalphORama** [microsoft/WSL#2477]
* Your Windows build number:  Microsoft Windows [Version 10.0.15063]

* What you're doing and what's happening:  `sudo apt-get update && sudo apt-get upgrade -y` downloads packages at less that 100 kB/s.
![image](https://user-images.githubusercontent.com/6566104/30195455-684b8568-9427-11e7-8846-56b57c12f316.png)

* What's wrong / what should be happening instead: `apt` should fetch packages much faster. My internet speed is 100 down / 10 up, on a bare-metal Linux installation or VirtualBox instance, `apt` is blazing fast. Packages are downloaded at upwards of 1024 kB/s.

* Strace of the failing command, if applicable:  n/a

**Comment by gravcat** [comment]
If there is nothing meaningful in the environment and you don't mind *completely destroying* it, try the following in an elevated command prompt:

`lxrun /uninstall /full /y`
`lxrun /install`

See if the problem persists if you completely redeploy the environment.

**Comment by sunilmut** [comment]
@RalphORama - Thanks for your post. We are aware that WSL networking perf is not at par with native Linux, but should not be unbearable. There are plans to improve the WSL n/w perf. I am not sure if uninstall/reinstalling will change much here for you.

**Comment by gravcat** [comment]
![network speeds wget](https://i.imgur.com/byeAF5a.png)

my network speeds are fine and maxing out my link speed on `Microsoft Windows [Version 10.0.15063]`

**Comment by therealkenc** [comment]
> We are aware that WSL networking perf is not at par with native Linux

Yeah WSL might not be on par with Real Linux, but we're talking phoronix-style quibbling not mass degradation. I get awful `apt-get` speeds like the OP sometimes too (measured kbs/sec on a mbps link), but best evidence seems to be it is on Canonical's end (or something in between). I get the same thing in my Ubunutu VM from time to time. I am not sure why this is, but I think (speculating) that sometimes you just get a redirect to an unlucky server.

@RalphORama
Best thing to do would be take Ubuntu out as a variable. Use speedtest from WSL CLI @gravcat suggest and then visit their [website](http://www.speedtest.net/) on Windows, and see if there is a serious gap in performance.
```
sudo apt install speedtest-cli
speedtest-cli
```
If you see an orders-of-magnitude difference between Windows and WSL, report back because that would be good to know; it is always possible there is something unique going on with your setup.

**Comment by Vaporeon** [comment]
I found that Windows Defender absolutely murders performance in WSL, especially tools like apt-get.
Try temporally disabling it and see if that improves it or not.

**Comment by heldchen** [comment]
> I found that Windows Defender absolutely murders performance in WSL, especially tools like apt-get.
Try temporally disabling it and see if that improves it or not.

this. Windows Defender really messes with I/O actions in WSL, not just with `apt-get`... adding scan exemptions does not seem to help much.

**Comment by therealkenc** [comment]
> this. Windows Defender really messes with I/O actions in WSL

It does. But keep in mind, in the OP this was the 'Get' phase, which is literally just a `curl` of a `.deb`. 46.6 _kilobytes_ per second on a 100Mbit/s downlink. Install phase I understand, sure. Windows Defender gets upset when thousands of little untrusted files start appearing from nowhere (from Defender's perspective) though its filesystem filter driver. But if it were Defender you should see the same slowdown on a bog standard curl from a fast server, and you don't; at least not to this degree. Couple that with the fact that turning Defender off doesn't help. So, something else. I could be wrong, because Defender is a ginormous black box that does who knows what, on or off. But I'm not seeing the causality here.

**Comment by RalphORama** [comment]
I reinstalled WSL as @gravcat suggested, and installed `speedtest-cli`. After running `speedtest-cli` and visiting the speedtest website, it appears WSL only uses 5% of my available bandwidth, maximum (1.6 Mbit/s vs 36 Mbit/s). I believe I have Windows Defender disabled, and I don't think I have any other programs that interfere with I/O on my machine. I use Malwarebytes as my antivirus, which doesn't run in the background. I'll try creating a firewall rule for `bash.exe` and report back.

![image](https://user-images.githubusercontent.com/6566104/30244975-e52b4e48-9598-11e7-80b0-94202b8a1b6b.png)
![image](https://user-images.githubusercontent.com/6566104/30244977-ee3d30aa-9598-11e7-9c9e-c253a99b48b6.png)

**Comment by vinibudd** [comment]
![image](https://user-images.githubusercontent.com/24849260/30517647-7aaacc02-9b3c-11e7-97e8-eb069e17722f.png)
-
![image](https://user-images.githubusercontent.com/24849260/30517653-8f0004ec-9b3c-11e7-8b74-99a44fa5e7b1.png)

**Comment by benhillis** [comment]
@therealkenc - Thanks for the repro steps.  I tracked this down to a directory rename failing due something that has a handle open to one of the directory's children. Unfortunately this is an NTFS limitation that I can't think of a workaround without NTFS supporting posix-style rename.  I know they were looking at that at some point, but I'm not sure of the status.

**Comment by therealkenc** [comment]
@benhillis - We talking about #640 here (not this one)? Either way, thanks for taking a peek.

**Comment by benhillis** [comment]
Oops yeah, got my threads crossed.

**Comment by DarkIlluminatus** [comment]
I am also experiencing this issue, did the solution suggested by @gravcat work for anyone?
I would hate to have to download the entire distro again at these below 100 kB/s speeds.

**Comment by leandrw** [comment]
Just turn off Windows Defender Real-time protection.

Add WSL folder and apt process in exclusion list will not work.

You should to completely disable real-time protection, reboot and will see how fast it will be

**Comment by DarkIlluminatus** [comment]
I no longer have this issue, it appears that I just lost the repo mirror
lotto when I did the dist-upgrade

**Comment by Garry050** [comment]
same here.
speedtest: both good speed (download: 70-90Mbps)
apt: always 30-300kb/s (ftp.jaist.ac.jp)

Using ESET Smart Security. no help disable anti-virus.
16299.214

**Comment by DarthSpock** [comment]
> Unfortunately this is an NTFS limitation that I can't think of a workaround without NTFS supporting posix-style rename. I know they were looking at that at some point, but I'm not sure of the status.

@benhillis Have you been able to get a status for this?

**Comment by benhillis** [comment]
@DarthSpock - unfortunately I think the posix-style rename work the ntfs team did doesn't handle the case where a directory has children with open handles. Adding @SvenGroot to confirm.

**Comment by sewashinobi** [comment]
I can confirm, this is slow as hell. For me the speed is 219 B/s, while my internet speed on windows is 24mbps

**Comment by mskd12** [comment]
I can confirm as well, the speed has dropped down significantly especially when downloading many things

**Comment by Wingmore** [comment]
Has anyone found a fix? getting speeds of 40 kB/s

**Comment by rondi74** [comment]
I'm having the same problem! >_<

**Comment by SvenGroot** [comment]
Ubuntu on WSL uses the Ubuntu's main archive mirror, which may not be the fastest option depending on your location. Could those who are having this issue try choosing a local mirror from https://launchpad.net/ubuntu/+archivemirrors, updating /etc/apt/sources.list to use that mirror instead of archive.ubuntu.com, and see if that improves your download speed?

**Comment by DarthSpock** [comment]
Running Synaptic on Kali goes rather fast. Faster than normal `apt` surprisingly. But even apt on Kali runs faster than Ubuntu. Which seems to point to agreement with what's said above.

**Comment by therealkenc** [comment]
You can confirm whether the variable is your Ubuntu mirror by testing against `speedtest.net` from within WSL with [speedtest-cli](https://github.com/sivel/speedtest-cli).
```
sudo apt install python-pip
pip install speedtest-cli   # installs in ~/.local/bin which is in your default path
speedtest-cli
```

**Comment by doomsbuster** [comment]
Turning off Windows Defender Realtime protection did the trick. All other options did not work.

**Comment by navneethc** [comment]
I'm having the same problems as well. Installed WSL only a month or so ago, and any time I try to install/update using apt-get I first get abysmal speeds (100s of Bps to KBps on a 75 Mbps connection)  and then everything comes to a screeching halt, at which point I kill it.

Windows 10.0.16299 (and Defender is turned off).

**Comment by bazald** [comment]
Just upgraded to build 17133.1 and this seems to be a new issue for me. Speed test results aren't bad. It seems like it's just taking forever to initialize each and every TCP connection. Pings are fine, as is actual bandwidth. Disabling Windows Defender Realtime protection has no effect.

**Comment by therealkenc** [comment]
> seems like it's just taking forever to initialize each and every TCP connection

I think (but can't prove because I don't have ipv6 with my ISP) the above issue might be a DNS lookup thing. Which is different than Sven's (very good) "distant mirror" hypothesis. There an issue# ref for the ipv6 DNS lookup problem but too lazy to look it up atm. Your _download speed_ with `apt` should be okay once the connection is made if that is the problem tho. One thing people can try to isolate is to [find](https://packages.ubuntu.com/) the `.deb` url of one of the "slow" gets that `apt` is fetching, and then just try getting it with `wget` to see how the speed compares. That is (to a first order) what `apt` is doing under the hood after all.

**Comment by bazald** [comment]
@therealkenc good guess! Disabling IPv6 resolves the issue. If it was a problem before the upgrade, it didn't seem as noticeable? I'm not sure why that would be, but it's interesting.

**Comment by therealkenc** [comment]
> If it was a problem before the upgrade, it didn't seem as noticeable?  I'm not sure why that would be, but it's interesting.

There were WSL improvements in some ipv6 scenarios since your last update. So what might be happening is your ipv6 use was just failing _outright_ before and falling back to ipv4 immediately. Now it is still failing, but "trying harder" (and taking longer). Couldn't tell you for sure because it would take a "before and after" `strace` compare, and that is not at all practical. But at least you are not alone; which is good in the scheme, since it is a known/reported problem as opposed to an unknown problem.

**Comment by ohnoimdead** [comment]
![image](https://user-images.githubusercontent.com/35269/39108783-d90cd19e-467e-11e8-890a-11e6ccc09cfd.png)

i realize this isn't exactly an apples-to-apples comparison, however, this is two somewhat similar ubuntu images (xenial in wsl and artful in vmware player) running on the same windows host at the same time. apt-get update called at as close to the same time as i could reasonably do manually. wsl takes nearly three minutes to complete compared to 2.5 seconds in a vm. happy to provide more information on my environment if that helps.

**Comment by nutcasev15** [comment]
Same problem on 1803 stable branch.
@bazald How did you disable ipv6? The sysctl method which one would use on normal ubuntu doesn't work here.

I however tried modifying the command at execution:

**Comment by bazald** [comment]
@nutcasev15 I disabled IPv6 for the entire distro by uncommenting "precedence ::ffff:0:0/96  100" in /etc/gai.conf.

**Comment by nutcasev15** [comment]
No change; confirms that my issues are not ipv6 or ipv4 specific.
Thanks Bazald

**Comment by therealkenc** [comment]
Using a slow server, which is demonstrated well by @ohnoimdead's post (one server is archive.ubuntu.com and the other is us.archive.ubuntu.com) is one issue. A separate issue is some folks having problems with ipv6 name resolution taking a long time. [A third issue, unrelated to this thread, is the install phase (versus get) being slow due to Defender among other things.]

**Comment by nutcasev15** [comment]
@therealkenc 
1) I changed my server to the local Indian version which works fine with Budgie in VM. Its not the server in my case.

2) My ISP does not have ipv6 support, as far as I know. So I have this problem on ipv4.

3) I have the slow install issue as well, slow download speed for packages too. Plus the apt-get update one.

Some logs to help clear things up:
ipv4 variant is with ``sudo apt-get update -o Acquire::ForceIPv4=true``
Default variant is without any modifications to any system configs & without forcing ipv4.
Both use the Indian server. Both hang at 28%.
[apt_log_ipv4.txt](https://github.com/Microsoft/WSL/files/2038058/apt_log_ipv4.txt)
[apt_log_default.txt](https://github.com/Microsoft/WSL/files/2038059/apt_log_default.txt)

**Comment by therealkenc** [comment]
> I changed my server to the local Indian version which works fine with Budgie in VM. Its not the server in my case.

How to proceed was explained above [here]([url](https://github.com/Microsoft/WSL/issues/2477#issuecomment-328224460)) and [here](https://github.com/Microsoft/WSL/issues/2477#issuecomment-371358156). Eliminate `apt` as a variable and see how your connection fares with speedtest.net under WSL versus Windows. If it is any consolation, I think there is something _else_ going on here _other_ than Sven's hypothesis (which may in fact be an issue for some folks). My ISP doesn't support ipv6 either and my apt get phase is unacceptably slow as well. 

"It's a process"

**Comment by nutcasev15** [comment]
& The Plot Unnecessarily Thickens:

http://www.speedtest.net/result/7339057179

``Retrieving speedtest.net configuration...
Testing from Excell Media Pvt (175.101.9.82)...
Retrieving speedtest.net server list...
Selecting best server based on ping...
Hosted by Excell Media (Hyderabad) [0.84 km]: 9.672 ms
Testing download speed................................................................................
Download: 18.86 Mbit/s
Testing upload speed................................................................................................
Upload: 33.82 Mbit/s
``
Upload in WSL is faster than Windows. WSL always had better upload, from the early days, but didn't think the speed would be close to double.
Meanwhile download suffers by the same factor in WSL.
The hell is happening?

**Comment by ohnoimdead** [comment]
@therealkenc interesting, i hadn't even noticed that apt in wsl was using the generic archive.ubuntu.com. updating to local us drastically reduced the time, however, it's still off by several seconds compared to vm (about 8 seconds now). oddly, an "npm install" (yes yes i know) seemed to take quite a bit longer on wsl vs vm for the same project about a month ago but now seems to be almost on par? ¯\_(ツ)_/¯

**Comment by nosferatu500** [comment]
If disable Windows diffender that issue is gone

**Comment by navneethc** [comment]
What is this black magic?! For the first time in 6 months I was able to successfully run update and upgrade. I haven't changed anything on my own, but between the last time I checked (some weeks ago) and today, there were some Windows updates (installed yesterday). Speeds weren't *that* impressive (started at 4000 kBps and eventually varied between 350-450 kBps), but considering what it used to be, it's a HUGE improvement.

However, now that this happened, I notice that 18.04 is available at the store. :D

**Comment by EOL007** [comment]
@SvenGroot your answer saved my ass! Changing to a local mirror feels like.. reborn. Thanks a lot!

**Comment by Neurrone** [comment]
I'm getting dialup-like speeds on a fibre connection, and none of the solutions here work for me.

* Disabling windows defender realtime protection
* Disabling IPV6
* Changing to a different mirror

**Comment by Neurrone** [comment]
```
Retrieving speedtest.net server list...
Selecting best server based on latency...
Testing download speed........................................
Download: 0.32 Mbit/s
Testing upload speed..................................................
Upload: 65.27 Mbit/s
```

Seems like a critical bug in downloading.

**Comment by mikevaleriano** [comment]
Had this issue when the distro was 16.04, trying again today with 18.04, the issue remains.

The slow network renders the whole product unusable, honestly. =/

**Comment by hrai** [comment]
Disabling the antivirus did it for me. I had symantec running.

**Comment by TimothyCole** [comment]
Any updates on this? I ran update/upgrade earlier today not realizing this was a thing now it's rendered my env completely useless.

I'm also not running an anti-virus (even Windows Defender)

**Comment by otonieru** [comment]
i also facing this issue

WSL only able to use a max bandwith around 200KB/s, while my connection is around 5MB/s (50Mbps)

and its not only for apt get, it also happen to git clone from google or github (or any) repo

disabling windows defender and use only iPv4 doesn't help.

**Comment by M4cs** [comment]
![image](https://user-images.githubusercontent.com/34947910/48959878-96ad1c80-ef36-11e8-8b90-e9d7c42a1bdb.png)

Im guessing this is just due to Kali servers being slow but my Windows Speedtest.net got these results:
<p align="center">
 <a href="http://www.speedtest.net/result/7824373557"><img src="http://www.speedtest.net/result/7824373557.png"/></a>
</p>

Same Upload speed but more than half of the download speed is cut.

Could windows system DNS settings effect this? I am routing through Cloudflare DNS.

**Comment by yuri-vashchenko** [comment]
The same here. tried everything suggested here, but nothing helps.
It has nothing to do with apt (I can live without often updates), but I cannot use subversion sever in local network because it takes ages to 'svn update', while the same 'svn update' completes almost instantly on hyper-v virtual machine

**Comment by M4cs** [comment]
I've been experiencing better speeds recently. Not sure what the cause of
it was.

**Comment by Po-wei** [comment]
Thanks @SvenGroot 
I finally solve my issue on extremely slow speed.
![image](https://user-images.githubusercontent.com/17596021/53790231-76dbba80-3f61-11e9-9993-f336fdf2cdb4.png)

Finally the speed is OK.
![image](https://user-images.githubusercontent.com/17596021/53805789-1612a900-3f86-11e9-9434-f79dbb76ead7.png)

**Comment by Idofriedman** [comment]
> > I found that Windows Defender absolutely murders performance in WSL, especially tools like apt-get.
> > Try temporally disabling it and see if that improves it or not.
> 
> this. Windows Defender really messes with I/O actions in WSL, not just with `apt-get`... adding scan exemptions does not seem to help much.

Worked for me as well

**Comment by Austin76016** [comment]
Building on what Sven said, here is a quick and easy command to change apt to automatically point to a mirror inside your country. `sed` -i -e 's/http:\/\/archive/mirror:\/\/mirrors/' -e 's/\/ubuntu\//\/mirrors.txt/' /etc/apt/sources.list`

**Comment by emincangencer** [comment]
Idk if this is related but with VM's if you use Nat the speed will be slower. Wsl uses Nat too as I know of. https://github.com/microsoft/WSL/issues/4150#issuecomment-504209723 . Try this for making it bridged.

**Comment by pcgeek86** [comment]
I'm also seeing very slow download performance (~400 KB/sec)for Windows Subsystem for Linux (WSL) on Windows 10 1903, using Ubuntu 18.04 LTS, when running `sudo apt-get upgrade --yes`.

**Comment by sirredbeard** [comment]
@pcgeek86 Have you tried using an alternative mirror for Ubuntu? https://launchpad.net/ubuntu/+archivemirrors

**Comment by therealkenc** [comment]
This thread has a few components, with different causes:

0. Conflating filesystem performance unrelated to networking. LZ for that is #2721 #981 (take your pick). Filesystem performance for apt updates has improved in WSL2. [For unrelated 9p fs performance in WSL2, LZ #4197]
1. Slow apt download phase due to a distant mirror. This much linux-behavior. Solution use a closer mirror.
2. Slow networking for reasons unknown and unlikely to be resolved in WSL1. This is partly for lack of being able to identify the variable causing the slowness to manifest for some, but not others. LZ for WSL2 is #4901

One can differentiate (1) from (2) with `speedtest-cli`.

**Comment by kevmo314** [comment]
I'm running into this issue with the Debian distribution. Running speedtest-cli indicates 80+ Mbit/s, however apt consistently downloads at 72.6 kB/s. I've tried disabling real-time protection and switching up the mirror to use a local debian package mirror to no avail, using [netselect-apt](https://linuxconfig.org/how-to-find-a-fastest-debian-linux-mirror-for-your-etc-apt-sources-list) to select the mirror.

Is there anything different with Debian under WSL? I can't imagine the archive is that slow for everyone...

**Comment by 01010000-kumar** [comment]
i'm using kali linux on WSL1, the `apt` mirrors are very slow.
i have a 5MB/s connection but the speed in WSL is just 25Kb/s
even the package installations are  #slow

is there a fix to this?

**Comment by sirredbeard** [comment]
@01010000-kumar 

To speed up apt downloads you can use official mirrors for your distro located closer to you*, use apt-fast, or even maintain your own package repo locally if you have a lot of machines.

If you can try WSL2 and WSL2 supports your workflow, you could try that.

*Azure, AWS, and other cloud services often provide their own apt repos which are also super fast.

**Comment by kewlgreen** [comment]
Disabling Windows Defender didn't seem to do much, but switching to a local mirror did the trick. I went from ~200 kB/s to ~1500 kB/s.

**Comment by aqeebimtiaz** [comment]
Following [THIS LINK](https://medium.com/@leandrocrs/speeding-up-wsl-i-o-up-than-5x-fast-saving-a-lot-of-battery-life-cpu-usage-c3537dd03c74), I created an exception on Windows Defender and it worked like a charm(!!!).

**TL;DR**

1. Get the UbuntuOnLinux (or your distro of choice) installation path: `%USERPROFILE%\AppData\Local\Packages` and lookup for something like **CanonicalGroupLimited.UbuntuonWindows_79rhkp1fndgsc**. Copy the path of that folder.
2. Open up your Windows Defender & add an exception of that folder. Follow:
`Settings > Update & Security > Windows Defender > Open Windows Defender Secutiry Central > Protection Against Viruses & Threats > Advanced Config… > Exclusions > Add or Remove > Add > Folder` and then select the folder. Or you can also paste that folder's path.

You should also check that article as it has other methods as well.

**Comment by josephexequiel** [comment]
Still having this issue on Ubuntu 20 WSL 1 and 2.

**Comment by noodlehaus** [comment]
same for me. WSL2 drops me back to 1.3~2 mbit/s, vs 200+ mbit/s on native windows.

**Comment by prakharbitspilani** [comment]
I tested by using wifi from mobile hotspots using 2 different network connections. Both having a speed of above 600kbps. However in one it works fine (even with windows defender turned on), and in other it doesn't work (even after disabling defender)

**Comment by noodlehaus** [comment]
> Following [THIS LINK](https://medium.com/@leandrocrs/speeding-up-wsl-i-o-up-than-5x-fast-saving-a-lot-of-battery-life-cpu-usage-c3537dd03c74), I created an exception on Windows Defender and it worked like a charm(!!!).
> 
> **TL;DR**
> 
> 1. Get the UbuntuOnLinux (or your distro of choice) installation path: `%USERPROFILE%\AppData\Local\Packages` and lookup for something like **CanonicalGroupLimited.UbuntuonWindows_79rhkp1fndgsc**. Copy the path of that folder.
> 2. Open up your Windows Defender & add an exception of that folder. Follow:
>    `Settings > Update & Security > Windows Defender > Open Windows Defender Secutiry Central > Protection Against Viruses & Threats > Advanced Config… > Exclusions > Add or Remove > Add > Folder` and then select the folder. Or you can also paste that folder's path.
> 
> You should also check that article as it has other methods as well.

Tried this a few days back, and this has certainly made things significantly better, though still not as fast as if i was outside WSL. But for code editing, the speed performance makes no difference, only when needing to download huge files.

**Comment by BeaudanBrown** [comment]
Haven't read all the responses to see how many people have tried this but the mirrors generated from [this](https://blog.zackad.dev/en/2018/01/18/how-to-select-fastest-ubuntu-repository-mirror.html) tool fixed the issue for me

**Comment by idamachmadfaizin** [comment]
Just turn off windows defender real-time protection.
or
Add WSL folder in windows defender exclusion list.
i'm using Ubuntu 20.04. and path is "C:\Users\%username%\AppData\Local\Packages\CanonicalGroupLimited.Ubuntu20.04onWindows_79rhkp1fndgsc".

**Comment by anvillasoto** [comment]
I fixed it by running this to PowerShell Admin:

New-NetFirewallRule -DisplayName "WSL" -Direction Inbound  -InterfaceAlias "vEthernet (WSL)"  -Action Allow

This basically opens the interface for WSL per this issue comment: #4585#issuecomment-610061194_

No turning of of windows defender necessary.

**Comment by pcgeek86** [comment]
@anvillasoto Wow, thank you! That actually works for me! I just tested it on an outdated instance of WSL. It's downloading packages at full speed now. 👍🏻

```powershell
New-NetFirewallRule -DisplayName WSL -Direction Inbound -InterfaceAlias "vEthernet (WSL)" -Action Allow
```

**Comment by NoelJacob** [comment]
> Building on what Sven said, here is a quick and easy command to change apt to automatically point to a mirror inside your country. `sed -i -e 's/http:\/\/archive/mirror:\/\/mirrors/' -e 's/\/ubuntu\//\/mirrors.txt/' /etc/apt/sources.list`

https://github.com/microsoft/WSL/issues/2477#issuecomment-541079233 by @Austin76016 worked, I guess.

Can anyone tell me this equalent for Kali-Linux? or how does this work?

**Comment by giosal** [comment]
Same issue here - all network related operations running very slowly on WSL 2 Ubuntu 20.04 running on Windows 10 x64 Insider Build 20241.
![image](https://user-images.githubusercontent.com/25568030/98589840-3273f680-22ce-11eb-8c4b-fe9a5e8d50b3.png)
Whereas from Chrome on Windows:
![image](https://user-images.githubusercontent.com/25568030/98589945-5fc0a480-22ce-11eb-81b1-bc8988815f17.png)
And another computer it's working perfectly (Windows 10 20H2):

**Comment by waqasaliBSSe-F17-58** [comment]
here is website you can use find internnet speed https://www.internetmeter.net/,if you want to find your ping them go to this link
https://www.internetmeter.net/ping.html if you want to find your jitter speed then simple click on this link https://www.internetmeter.net/jitter.html
best internet speed finder website https://www.internetmeter.net/

**Comment by waqasaliBSSe-F17-58** [comment]
here is website you can use find internnet speed https://www.internetmeter.net/,if you want to find your ping them go to this link
https://www.internetmeter.net/ping.html if you want to find your jitter speed then simple click on this link https://www.internetmeter.net/jitter.html
best internet speed finder website https://www.internetmeter.net/

**Comment by chuanqisun** [comment]
Same issue. WSL 2, Ubuntu 20.04 LTS on Windows 10 Pro 19041.630

It's definitely not caused by a slow mirror.

![image](https://user-images.githubusercontent.com/1895289/100131129-7d4a4e00-2e38-11eb-807c-5b21a79a4308.png)

**Comment by Aymane11** [comment]
> 
> 
> Following [THIS LINK](https://medium.com/@leandrocrs/speeding-up-wsl-i-o-up-than-5x-fast-saving-a-lot-of-battery-life-cpu-usage-c3537dd03c74), I created an exception on Windows Defender and it worked like a charm(!!!).
> 
> **TL;DR**
> 
>     1. Get the UbuntuOnLinux (or your distro of choice) installation path: `%USERPROFILE%\AppData\Local\Packages` and lookup for something like **CanonicalGroupLimited.UbuntuonWindows_79rhkp1fndgsc**. Copy the path of that folder.
> 
>     2. Open up your Windows Defender & add an exception of that folder. Follow:
>        `Settings > Update & Security > Windows Defender > Open Windows Defender Secutiry Central > Protection Against Viruses & Threats > Advanced Config… > Exclusions > Add or Remove > Add > Folder` and then select the folder. Or you can also paste that folder's path.
> 
> 
> You should also check that article as it has other methods as well.

Excluding the `CanonicalGroupLimited.....` folder from my antivirus solved the problem for me.

**Comment by danieliser** [comment]
@Aymane11 I actually have a PowerShell script I use to do this automatically for all WSL instances.

```
############
# This script will add your WSL environments to the Windows Defender exclusion list so that
# realtime protection does not have an adverse effect on performance.
#
# You should be aware that this could make your system less secure. Use at your own risk.
# Note: This should be run from an administrative PowerShell prompt
############

# Find registered WSL environments
$wslPaths = (Get-ChildItem HKCU:\Software\Microsoft\Windows\CurrentVersion\Lxss | ForEach-Object { Get-ItemProperty $_.PSPath}).BasePath

# Get the current Windows Defender exclusion paths
$currentExclusions = $(Get-MpPreference).ExclusionPath
if (!$currentExclusions) {
  $currentExclusions = ''
}

# Find the WSL paths that are not excluded
$exclusionsToAdd = ((Compare-Object $wslPaths $currentExclusions) | Where-Object SideIndicator -eq "<=").InputObject

# List of paths inside the Linux distro to exclude (https://github.com/Microsoft/WSL/issues/1932#issuecomment-407855346)
$dirs = @("\bin", "\sbin", "\usr\bin", "\usr\sbin", "\usr\local\bin", "\usr\local\go\bin")

# Add the missing entries to Windows Defender
if ($exclusionsToAdd.Length -gt 0) {
  $exclusionsToAdd | ForEach-Object { 
    
    # Exclude paths from the root of the WSL install
    Add-MpPreference -ExclusionPath $_
    Write-Output "Added exclusion for $_"

    # Exclude processes contained inside WSL
    $rootfs = $_ + "\rootfs"
    $dirs | ForEach-Object {
        $exclusion = $rootfs + $_ + "\*"
        Add-MpPreference -ExclusionProcess $exclusion
        Write-Output "Added exclusion for $exclusion"
    }
  }
}
```

**Comment by danieliser** [comment]
Hope this part helps everyone, though I had the above for I/O performance improvements, it had no impact on connection speeds.

However what I did find, and my tests are below, is that disabling Windows Defender Firewall (Internet rules) does resolve it.

Baseline: Speedtest from Windows)
![image](https://user-images.githubusercontent.com/847818/106381692-94ea0980-6388-11eb-95e0-9f61ab03551a.png)

WSL Before & After disabling Firewall
![image](https://user-images.githubusercontent.com/847818/106381671-6d933c80-6388-11eb-91c8-962306701494.png)

Next I need to figure out what needs to be unblocked to ensure this doesn't happen without leaving firewall off entirely.

Any thoughts on specifically what to unblock? Can't imagine port based rules will do. Not sure which application you could select to unblock either.

**Comment by danieliser** [comment]
I have narrowed the docker issue as well now. Seems to be a bottleneck due to using a WiFi connection.

Switching to a wired connection and WSL2 -> Docker CE containers for tensorflow with GPU support will now download using the entire connection.

No idea why that is. Haven't been able to find any info on why docker has issues with wifi.

**Comment by SanderElias** [comment]
I'm wired to a 600Mb connection. This is the only thing that is slow:
![image](https://user-images.githubusercontent.com/1249083/122542580-24881000-d02b-11eb-86f0-cbcf5fb0cbba.png)

I already switched connections, removed code-insiders completely, and reinstalled them.
The first reinstallation goes quick then, but with every upgrade I'm back to slowness?

**Comment by Arian8j2** [comment]
@SvenGroot Solution works for me, Also remember to change security mirrors too, So completed command will be like:
`mirror="us.archive.ubuntu.com" && sudo sed -i "s/archive.ubuntu.com/$mirror/; s/security.ubuntu.com/$mirror/" /etc/apt/sources.list`, Also not sure that is it good idea to change security mirrors or not.

**Comment by emadha** [comment]
i copied the `/etc/resolve.conf` from ubuntu to kali's resolv.conf and it solved the problem

# [network]
# generateResolvConf = false
nameserver 192.x.x.184
nameserver xxxx:0:0:ffff::1
nameserver xxxx:0:0:ffff::2

this solved my kali issue, I had to `wsl --showdown` from powershell and restart kali for this to work.

**Comment by jhagege** [comment]
I realized the issue for me was that antivirus ESET was blocking the traffic for WSL Ubuntu. 
Unblocked by going to Setup --> Network --> Troubleshooting Wizard --> Unblock.

**Comment by craigloewen-msft** [comment]
Hi folks, we have put out a new update that aims to address networking issues in WSL. In your `.wslconfig` file you can set `experimental.networkingMode=mirrored`, as well as some other key settings that should improve your network compatibility! Please try them out and let us know what you think.

More info on this release and the changes can be [found here in the blog post](https://devblogs.microsoft.com/commandline/windows-subsystem-for-linux-september-2023-update/).

Please note: You need to be on a Windows Insiders version to use the new networking settings (Any channel of Windows Insiders will do, including release preview). If you see the "These are not supported" messages it means that your current Windows version doesn't have support, and you will need to upgrade. These features will eventually be coming to Windows 11 22H2.

**Comment by meydiwahendra** [comment]
if you're an asian, its easy to fix this issues ..
# sudo sed -i "s/archive.ubuntu.com/sg.archive.ubuntu.com/" /etc/apt/sources.list
# sudo apt update
# see the different !!!

**Comment by DrorDvash** [comment]
> I realized the issue for me was that antivirus ESET was blocking the traffic for WSL Ubuntu. Unblocked by going to Setup --> Network --> Troubleshooting Wizard --> Unblock.

Thanks! helped me too with ESET Internet Security

**Comment by OneBlue** [comment]
Hi ! Please try the [latest networking features](https://learn.microsoft.com/en-us/windows/wsl/networking#dns-tunneling) that we've added in WSL. Those should greatly improve performance. 

If the issue still remains, please reopen this issue.

**Comment by bukowa** [comment]
Changing nameservers

> Hi ! Please try the [latest networking features](https://learn.microsoft.com/en-us/windows/wsl/networking#dns-tunneling) that we've added in WSL. Those should greatly improve performance.
> 
> If the issue still remains, please reopen this issue.

Doesn't work for me.
The only thing that works is to change dns in `/etc/resolv.conf`
